pinax/ratings/tests/test_views.py

```diff
@@ -0,0 +1,112 @@
+from django.contrib.contenttypes.models import ContentType
+
+from .models import Car
+from .test import TestCase
+
+
+class RateViewTest(TestCase):
+    """
+    tests for the RateView
+    """
+
+    def setUp(self):
+        self.test_user = self.make_user(username="test_user")
+        self.forester = Car.objects.create(name="Subaru Forester")
+
+    def test_post_rating_returns_404(self):
+        """
+        Ensure view returns 404 if object is not found. This should
+        typically happen when one of the two kwargs i.e content_type_id
+        and object_id have invalid data
+        """
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=400,
+                object_id=1,
+                data={
+                    "rating": 3,
+                    "category": "handling"
+                },
+            )
+            self.response_404(response)
+
+    def test_post_rating_with_invalid_category(self):
+        """
+        Ensure view returns 403 if an invalid category choice
+        is provided in the POST data. category choices are added
+        to the PINAX_RATINGS_CATEGORY_CHOICES setting in settings.py
+        """
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 4,
+                    "category": "non-existing-category"
+                },
+            )
+            self.response_403(response)
+
+    def test_post_rating_greater_than_range(self):
+        """
+        Ensure view returns 403 if the rating is greater than the
+        NUM_OF_RATINGS setting added to the PINAX_RATINGS_NUM_OF_RATINGS setting
+        in settings.py
+        """
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 4,
```

Comment:
Constants like `4` in this test often cause trouble. What happens if another developer changes the setting in runtests.py to 5? This test breaks, but not for a good reason.

Rather than use a constant, import and reference the external value. In this case I suggest adding `from django.conf import settings` up top, and

```
    "rating": settings.PINAX_RATINGS_NUM_OF_RATINGS + 1
```

in the test. The test is clearer as well, showing _why_ the failure occurs... a value 1 greater than PINAX_RATINGS_NUM_OF_RATINGS is too much.

Reply:
@grahamu constant has been added